torchgeo/datasets/geo.py

```diff
@@ -525,34 +532,89 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
-                    if match:
-                        if 'band' in match.groupdict():
-                            start = match.start('band')
-                            end = match.end('band')
-                            filename = filename[:start] + band + filename[end:]
-                    filepath = os.path.join(directory, filename)
-                    band_filepaths.append(filepath)
-                data_list.append(self._merge_files(band_filepaths, query))
-            data = torch.cat(data_list)
+        if self.time_series:
+            # Group by unique datetime intervals to support timeseries
+            grouped_index = index.groupby(index.index)
+            time_steps = []
+            dates = []
+
+            for datetime_interval, group in grouped_index:
+                # Extract the center date of the interval
+                center_date = datetime_interval.mid
+                dates.append(center_date)
+
+                if self.separate_files:
+                    # For separate files, we need to collect files by band for this time step
+                    band_data: list[Tensor] = []
+                    filename_regex = re.compile(self.filename_regex, re.VERBOSE)
+
+                    for band in self.bands:
+                        # For each band, find all files for this time interval
+                        band_filepaths = []
+                        for filepath in group.filepath:
+                            filename = os.path.basename(filepath)
+                            directory = os.path.dirname(filepath)
+                            match = re.match(filename_regex, filename)
+                            if match:
+                                if 'band' in match.groupdict():
+                                    start = match.start('band')
+                                    end = match.end('band')
+                                    filename = filename[:start] + band + filename[end:]
+                            filepath = os.path.join(directory, filename)
+                            band_filepaths.append(filepath)
+
+                        # Merge files for this band at this time step
+                        band_data.append(self._merge_files(band_filepaths, query))
+
+                    # Concatenate all bands for this time step
+                    time_step_data = torch.cat(band_data, dim=0)
+                else:
+                    # For non-separate files, merge all files for this time step
+                    time_step_data = self._merge_files(
+                        group.filepath, query, self.band_indexes
+                    )
+
+                time_steps.append(time_step_data)
+
+            # Stack along time dimension to create [T,C,H,W]
+            data = torch.stack(time_steps, dim=0)
+
         else:
-            data = self._merge_files(index.filepath, query, self.band_indexes)
+            # Original non-timeseries behavior
```

Comment:
Let's remove this comment, "original" won't make sense in a few years.

Reply:
Done